### DISCUSSION TITLE
Advice on preparing XML-dependent library for NativeAOT

### DISCUSSION THREAD

**Discussion by karakasa** [dotnet/runtime#89664] [Q&A] [answered]
I've been preparing [NPOI](https://github.com/nissl-lab/npoi), an xls(x) read-and-write library heavily relying on XML (de)serialization, for NativeAOT.

1. Is there general advice on how to prepare `rd.xml` for such a XML-dependent library?
2. Are there source generators or other AOT-compatibile ways to do object deepcopy?
3. Is it possible to do, sort of, unit tests on AOTed versions now?

**Comment by hez2010** [comment]
rd.xml is not intended to be used by libraries. Instead, annotate your libraries with DynamicallyAccessedMembersAttribute, etc.

**Comment by jkotas** [comment]
> heavily relying on XML (de)serialization,

Some libraries are not going to be fully compatible with native AOT. It is likely the case of System.Xml.Serialization.

**Reply by karakasa** [reply]
I haven't tested it thoroughly but it looks like the library (and the underlying XML deserialization) works OK once I resolved obvious Reflection-related issues. Hopefully unit tests can work with NativeAOT with less effort.

**Reply by jkotas** [reply, marked as answer]
[Annotating the library](https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/prepare-libraries-for-trimming) and getting to zero trim and AOT warnings is the only fully reliable way to make libraries work well with trimming and AOT.

It is not possible to find all trimming and AOT issues using unit test.

**Comment by idan-h** [comment]
What I did was write a custom source generator that parses the xml. Worked for my case.
Not sure why microsoft doesn't drop even a minimal xml source generator with common cases, was not very hard to implement..
